src/components/measurements/measurementsD3.js

```diff
@@ -0,0 +1,338 @@
+import { extent, groups, mean, deviation } from "d3-array";
+import { axisBottom, axisLeft } from "d3-axis";
+import { scaleLinear } from "d3-scale";
+import { select, event as d3event } from "d3-selection";
+import { symbol, symbolDiamond } from "d3-shape";
+import { orderBy } from "lodash";
+
+/* C O N S T A N T S */
+const layout = {
+  leftPadding: 120,
+  rightPadding: 30,
+  topPadding: 20,
+  bottomPadding: 50,
+  subplotHeight: 100,
+  subplotPadding: 10,
+  circleRadius: 3,
+  circleHoverRadius: 5,
+  thresholdStrokeWidth: 2,
+  thresholdStroke: "#DDD",
+  subplotFill: "#adb1b3",
+  subplotFillOpacity: "0.15",
+  diamondSize: 25,
+  standardDeviationStroke: 2,
+  overallMeanColor: "#000"
+};
+
+const classes = {
+  xAxis: "measurementXAxis",
+  yAxis: "measurementYAxis",
+  threshold: "measurementThreshold",
+  subplot: "measurementSubplot",
+  subplotBackground: "measurementSubplotBackground",
+  rawMeasurements: "rawMeasurements",
+  rawMeasurementsGroup: "rawMeasurementsGroup",
+  overallMean: "measurementsOverallMean",
+  colorMean: "measurementsColorMean",
+  mean: "mean",
+  standardDeviation: "standardDeviation"
+};
+
+export const svgContainerDOMId = "measurementsSVGContainer";
+const getMeasurementDOMId = (measurement) => `meaurement_${measurement.measurementId}`;
+const domIdRegex = new RegExp("[^\\w\\-_]+", "gi");
+const getSubplotDOMId = (groupingValue) => `measurment_subplot_${groupingValue.replace(domIdRegex, "_")}`;
```

Comment:
Replacements like this one that collapse a) all forbidden chars into a single replacement char and/or b) runs of forbidden chars into a single replacement char can produce colliding values in edge cases. Since grouping values here are user-provided and not fixed values, it'd be good to make this more robust.

One way is to generate completely synthetic ids for each `groupingValue`, e.g. `measurement_subplot_1`, as is done for measurements. This requires a bit more bookkeeping.

Another way is to choose some CSS selector safe encoding, like base64, instead of replacing chars (base64 in JS [requires a little dancing](https://gist.github.com/tsibley/618f119d1cd3bf3da885170403ebfba7)), e.g. `measurement_subplot_Zm9v4pmlYmFy`.

Stepping back though, it's worth nothing that DOM attributes, including id, support arbitrary strings. So the issue is primarily with supporting arbitrary ids for the CSS selector syntax further down:

```js
const svg = select(ref);
// …
const subplot = svg.select(`#${getSubplotDOMId(groupingValue)}`);
```

If the ids aren't needed for other purposes, then instead of using a CSS selector, this could change to:

```
const subplot = ref.getElementById(getSubplotDOMId(groupingValue));
```

and have `getSubplotDOMId()` stop doing any replacement.

(Also, minor typo of no consequence: `measurment_subplot_` → `measurement_subplot_`)

Reply:
Resolved in 5921b96. I went with the synthetic ids since I had to do the extra bookkeeping to keep stable ordering of subplots anyways.